internal/http/services/owncloud/ocdav/tpc.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	sublog := appctx.GetLogger(ctx).With().Str("src", src).Str("dst", dst).Logger()
+	sublog.Debug().Str("overwrite", overwrite).Str("depth", depth).Msg("TPC Pull")
+
+	overwrite = strings.ToUpper(overwrite)
+	if overwrite == "" {
+		overwrite = "T"
+	}
+
+	if overwrite != "T" && overwrite != "F" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if depth != "infinity" && depth != "0" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// get Gateway client
+	client, err := s.getClient()
+	if err != nil {
+		sublog.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// check if destination exists
+	ref := &provider.Reference{Path: dst}
+	dstStatReq := &provider.StatRequest{Ref: ref}
+	dstStatRes, err := client.Stat(ctx, dstStatReq)
+	if err != nil {
+		sublog.Error().Err(err).Msg("error sending grpc stat request")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if dstStatRes.Status.Code != rpc.Code_CODE_OK && dstStatRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+		HandleErrorStatus(&sublog, w, dstStatRes.Status)
+		return
+	}
+
+	if overwrite == "F" {
+		sublog.Warn().Str("overwrite", overwrite).Msg("Destination already exists")
+		w.WriteHeader(http.StatusPreconditionFailed) // 412, see https://tools.ietf.org/html/rfc4918#section-9.8.5
+		return
+	}
+
+	w.WriteHeader(http.StatusAccepted)
+
+	err = s.performHTTPPull(ctx, client, r, w, depth == "infinity", ns)
+	if err != nil {
+		sublog.Error().Err(err).Str("depth", depth).Msg("error descending directory")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	fmt.Fprintf(w, "success: Created")
+}
+
+func (s *svc) performHTTPPull(ctx context.Context, client gateway.GatewayAPIClient, r *http.Request, w http.ResponseWriter, recurse bool, ns string) error {
+	src := r.Header.Get("Source")
+	dst := path.Join(ns, r.URL.Path)
+	size := 1024
+	log := appctx.GetLogger(ctx)
+	log.Debug().Str("src", src).Str("dst", dst).Msg("Performing HTTP Pull")
+
+	// get upload url
+	uReq := &provider.InitiateFileUploadRequest{
+		Ref: &provider.Reference{Path: dst},
+		Opaque: &typespb.Opaque{
+			Map: map[string]*typespb.OpaqueEntry{
+				"Upload-Length": {
+					Decoder: "plain",
+					// TODO: handle case where size is not known in advance
+					Value: []byte(fmt.Sprintf("%d", size)),
+				},
+			},
+		},
+	}
+	uRes, err := client.InitiateFileUpload(ctx, uReq)
+	if err != nil {
+		return err
+	}
+
+	if uRes.Status.Code != rpc.Code_CODE_OK {
+		return fmt.Errorf("status code %d", uRes.Status.Code)
+	}
+
+	var uploadEP, uploadToken string
+	for _, p := range uRes.Protocols {
+		if p.Protocol == "simple" {
+			uploadEP, uploadToken = p.UploadEndpoint, p.Token
+		}
+	}
+
+	// get http client for remote
+	httpClient := &http.Client{}
+
+	req, err := http.NewRequest("GET", src, nil)
+	if err != nil {
+		return err
+	}
+
+	// add authentication header
+	bearerHeader := r.Header.Get("TransferHeaderAuthorization")
```

Comment:
Add it as a constant here https://github.com/cs3org/reva/blob/master/internal/http/services/owncloud/ocdav/webdav.go#L57